src/biokbase/narrative/jobs/job.py

```diff
@@ -225,22 +293,28 @@ def _augment_ee2_state(self, state):
             output_state[arg] = getattr(self, arg)
         return output_state
 
-    def state(self):
+    def state(self, force_refresh=False):
         """
         Queries the job service to see the state of the current job.
         """
 
-        if self.terminal_state:
-            return self._last_state
+        if self.was_terminal and not force_refresh:
+            state = copy.deepcopy(self._last_state)
+        else:
+            try:
+                state = clients.get("execution_engine2").check_job(
+                    {"job_id": self.job_id, "exclude_fields": EXCLUDED_JOB_STATE_FIELDS}
+                )
+                state = self.update_state(state)
 
-        try:
-            state = clients.get("execution_engine2").check_job(
-                {"job_id": self.job_id, "exclude_fields": EXCLUDED_JOB_STATE_FIELDS}
-            )
-            return self.update_state(state)
+            except Exception as e:
+                raise Exception(f"Unable to fetch info for job {self.job_id} - {e}")
 
-        except Exception as e:
-            raise Exception(f"Unable to fetch info for job {self.job_id} - {e}")
+        for field in EXCLUDED_JOB_STATE_FIELDS:
+            if field in state and field != "job_input":
+                del state[field]
```

Comment:
this might be better in the `update_state` method to ensure that the keys in that list never get saved

Reply:
I think such goalkeeping makes me a little googly-eyed, as there are a number of entrypoints for the EE2 state: there's `from_attributes` and `from_state`, which then I put in `self._const_state` and `self._last_state`; and there's `update_state` and `state()`. Psychologically it also seems to make more sense to just accept everything but prune/augment when emitting, which only technically happens at `state()` and `output_state`.

Reply:
I am assuming not an astronomical number of `Job` instances and a comfortable amount of computing resources